[Experimental] Product Filters: add block spacing setting

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Partial closes #52733.

This PR:
- Updates the default heading structure for Product Filters, with H3 for the parent block heading, and H4 for each filter block heading.
- Adds block spacing setting to the Product Filters block.
- Fixes some issues with List and Chips that use an undefined argument in the Save function.
- Restores the default render routine for the Product Filters block to let WP decorate the block classes and style.

Note:
- I didn't add the E2E test as I don't this is a critical flow. Unit test for new utility is pending.
- The block utilizes the saved content and uses WP HTML Tag Processor to extract the class and style from it. So if there is a Product Filters on the page, you will get the block validation notice. As the Product Filters are still experimental, it's not a problem. When we release the block, we will need to use Deprecation API to avoid that notice.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1. Add the `Product Filters` block to the `Product Catalog` template.
2. Open the inspector setting > Styles.
3. See block spacing setting.
4. Changing the setting, see the editor preview updates.
5. Save and go to the frontend.
6. See the block spacing applied.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dinhtungdu committed:
fix: update heading levels

dinhtungdu committed:
fix: remove heading margin

dinhtungdu committed:
fix: removable chip layout style

dinhtungdu committed:
fix: remove undefined style argument

dinhtungdu committed:
fix: use default render routine to let WP decorate the block style and classes

dinhtungdu committed:
feat: add utility to process Product Filters style

dinhtungdu committed:
feat: add block spacing support

dinhtungdu committed:
chore: changelog

roykho reviewed (changes requested):
Looking good so far however I think the default gap spacing is too large. There is too much space in between.

thealexandrelara reviewed (changes requested):
Nice, it looks much better in the editor, but I see that the spacing is not being applied correctly on the frontend.

✅ Scenario 1: The spacing between blocks is correct in the Editor.
- Conditions:
    - A merchant visits the Product Catalog template.
    - The merchant adds the Product Filters (Experimental) block.
- Expected Outcome: The spacing between blocks is correct.
- Summary: Scenario passed as expected. No issues were found.

---

⚠️ Scenario 2: The spacing between blocks is correct on the frontend.
- Conditions:
    - A merchant visits the Product Catalog template.
    - The merchant adds the Product Filters (Experimental) block.
- Expected Outcome: The spacing between blocks is correct.
- Summary: The spacing is not being applied correctly on the frontend. It looks like the CSS variable does not exist:

![CleanShot 2024-11-13 at 14 27 15@2x](https://github.com/user-attachments/assets/3b19314a-b148-4f9d-89c2-3ab09e0a60ea)

Here is the result:

![CleanShot 2024-11-13 at 14 28 07@2x](https://github.com/user-attachments/assets/7185f28f-0a16-4926-a0b8-6045a07d96a0)

dinhtungdu committed:
fix: smaller default gap

dinhtungdu commented:
@roykho I fixed the default gap in https://github.com/woocommerce/woocommerce/pull/52750/commits/9da65c0d2f0458147fff6a60c504c693c5915b4a.

> ⚠️ Scenario 2: The spacing between blocks is correct on the frontend.

@thealexandrelara did you set a custom block spacing setting? If no, then the block will use the fallback value, which is `inherit` at the time you test (and changed to `--wp--style--block-gap` in https://github.com/woocommerce/woocommerce/pull/52750/commits/9da65c0d2f0458147fff6a60c504c693c5915b4a)

thealexandrelara reviewed (approved):
Thank you, Tung! I believe your last commit did the trick, now everything is working as expected. 🚀 

✅ Scenario 1: The spacing between blocks is correct in the Editor.
- Conditions:
    - A merchant visits the Product Catalog template.
    - The merchant adds the Product Filters (Experimental) block.
- Expected Outcome: The spacing between blocks is correct.
- Summary: Scenario passed as expected. No issues were found.

---

✅ Scenario 2: The spacing between blocks is correct on the frontend.
- Conditions:
    - A merchant visits the Product Catalog template.
    - The merchant adds the Product Filters (Experimental) block.
- Expected Outcome: The spacing between blocks is correct.
- Summary: Scenario passed as expected. No issues were found.

roykho reviewed (approved):
Thanks for fixing up the default gap. Looking good now.

✅ Default gap is visually balanced
✅ Increasing/decreasing the gap is reflected in the editor and frontend.

dinhtungdu committed:
chore: remove unused import
